### PR TITLE
Check validity of oio_url_s instance

### PR DESCRIPTION
**Pull request by murlock**
Only core, meta2 and proxy are now protected (I hope, please correct me if I miss something).

Path and version are always checked with oio_url_check.
Namespace is only checked if current one is passed as argument.

### CODE REVIEW

File: meta2v2/meta2_filters_extract.c [hunk 1]
```diff
@@ -77,6 +77,12 @@ meta2_filter_extract_header_url(struct gridd_filter_ctx_s *ctx,
 {
 	TRACE_FILTER();
 	struct oio_url_s *url = metautils_message_extract_url (reply->request);
+    if (!oio_url_check(url)) {
+		meta2_filter_ctx_set_error(ctx, NEWERROR(CODE_BAD_REQUEST,
+					"Invalid request, Empty / Invalid body"));
```
Comment (fvennetier) [inline meta2v2/meta2_filters_extract.c]: The message should tell the problem comes from the content/container URL.

File: proxy/metacd_http.c [hunk 2]
```diff
@@ -238,7 +238,9 @@ handler_action (struct http_request_s *rq, struct http_reply_ctx_s *rp)
 		if (oio_url_has(args.url, OIOURL_VERSION) && !oio_str_is_number(
 					oio_url_get(args.url, OIOURL_VERSION), &args.version))
 			rc = _reply_format_error(&args, BADREQ("Invalid version"));
-		else {
+		else if (!oio_url_check(url)) {
+			rc = _reply_format_error(&args, BADREQ("Invalid parameters"));
```
Comment (fvennetier) [inline proxy/metacd_http.c]: Same here, "parameters" is too vague.

File: core/url.c [hunk 3]
```diff
@@ -506,3 +506,8 @@ oio_url_to_json (GString *out, struct oio_url_s *u)
 	}
 }
 
+gboolean
+oio_url_check(const struct oio_url_s *u)
+{
+    return !u->path || g_utf8_validate(u->path, -1, NULL);
```
Comment (jfsmig) [inline core/url.c]: You could also validate the format of the version, e.g. with `oio_str_is_number()`.

File: proxy/metacd_http.c [hunk 4]
```diff
@@ -238,7 +238,9 @@ handler_action (struct http_request_s *rq, struct http_reply_ctx_s *rp)
 		if (oio_url_has(args.url, OIOURL_VERSION) && !oio_str_is_number(
```
Comment (jfsmig) [inline proxy/metacd_http.c]: Checking the version could be part of the `oio_url_check()` function.

File: proxy/metacd_http.c [hunk 5]
```diff
@@ -238,7 +238,9 @@ handler_action (struct http_request_s *rq, struct http_reply_ctx_s *rp)
...
 					oio_url_get(args.url, OIOURL_VERSION), &args.version))
 			rc = _reply_format_error(&args, BADREQ("Invalid version"));
-		else {
```
Comment (jfsmig) [inline proxy/metacd_http.c]: You could also check the NS (if present) is managed by the current `oio-proxy`, as in https://github.com/open-io/oio-sds/blob/master/proxy/lb_actions.c#L229

File: meta1v2/meta1_gridd_dispatcher.c [hunk 6]
```diff
@@ -46,13 +46,29 @@ static GByteArray *encode_and_clean(GByteArray* (*e)(gchar**), gchar **pv) {
...
 /* -------------------------------------------------------------------------- */
 
 static gboolean
 meta1_dispatch_v2_USERCREATE(struct gridd_reply_ctx_s *reply,
 		struct meta1_backend_s *m1, gpointer ignored UNUSED)
 {
-	struct oio_url_s *url = metautils_message_extract_url (reply->request);
+	struct oio_url_s *url = _extract_url(reply);
+	if (!url) {
+		return TRUE;
```
Comment (jfsmig) [inline meta1v2/meta1_gridd_dispatcher.c]: It is highly recommended either to reply something to the client or to abruptly close the connection with a `return FALSE`.

Comment (murlock) [inline meta1v2/meta1_gridd_dispatcher.c]: I bet that my function _extract_url is misnamed because a reply is done with BADREQ

**Pull request merged**